src/levanter/models/factorized_llama.py

```diff
@@ -0,0 +1,627 @@
+            rope_scaling=self.rope_scaling,
+            factor_dim=self.factor_dim,
+            vocab_size=vocab_size,
+            **config_overrides,
+        )
+
+    @property
+    def model_type(cls) -> Type["FactorizedLlamaLMHeadModel"]:
+        return FactorizedLlamaLMHeadModel
+
+
+def low_rank_approximation(matrix, rank):
+    """
+    Approximates the input matrix using Singular Value Decomposition (SVD) to a lower rank.
+
+    Args:
+        matrix (jax.numpy.ndarray): Input matrix of shape (N, M)
+        rank (int): Desired rank of the approximation (H)
+
+    Returns:
+        jax.numpy.ndarray, jax.numpy.ndarray: Two matrices of shape (N, H) and (H, M)
+    """
+    from jax.numpy.linalg import svd
+
+    # Perform SVD
+    U, S, Vh = svd(matrix, full_matrices=False)
+
+    S = S[..., :rank]
+    # Truncate U, S, and Vh to the desired rank
+    U_truncated = U[..., :rank]
+    if len(S.shape) == 1:
+        S_truncated = jnp.diag(S)
+    else:
+        S_truncated = jax.vmap(jnp.diag, 0)(S[..., :rank])
+
+    Vh_truncated = Vh[..., :rank, :]  # Note: Vh is already the conjugate transpose of V
+
+    # Reconstruct the low-rank approximation
+    down_proj = U_truncated @ S_truncated  # Shape (N, H)
+    up_proj = Vh_truncated  # Shape (H, M)
+
+    return down_proj, up_proj
+
+
+class FactorizedLinear(StateDictSerializationMixin, eqx.Module):
+    """Factorized Linear Layer"""
+
+    down_proj: hnn.Linear
+    up_proj: hnn.Linear
+    out_first: bool = eqx.static_field()
+
+    @staticmethod
+    def init(
+        Out: Axis, In: Axis, Hidden: Axis, *, key, use_bias: bool = False, out_first: bool = False
+    ) -> "FactorizedLinear":
+        assert Hidden.size <= np.prod([a.size for a in ensure_tuple(Out)]), (
+            "Hidden size must be less than or equal to output size.",
+            Hidden,
+            Out,
+        )
+
+        k_down, k_up = jrandom.split(key, 2)
+        down_proj = hnn.Linear.init(Out=Hidden, In=In, key=k_up, use_bias=use_bias)
+        up_proj = hnn.Linear.init(Out=Out, In=Hidden, key=k_down, use_bias=use_bias)
+        return FactorizedLinear(down_proj, up_proj, out_first=out_first)
+
+    def __call__(self, x: NamedArray, *, key=None) -> NamedArray:
+        k_down, k_up = maybe_rng_split(key, 2)
+        return self.up_proj(self.down_proj(x, key=k_up), key=k_down)
+
+    def from_state_dict(self, state_dict: StateDict, prefix: Optional[str] = None):
+        d = state_dict.copy()
+
+        # Initial factorized linear with the SVD approximation of the original input matrix.
+        weights = state_dict[prefix + ".weight"]
+        down_proj, up_proj = low_rank_approximation(weights, self.down_proj.Out.size)
+        if self.out_first:
+            d[prefix + ".down_proj.weight"] = up_proj
+            d[prefix + ".up_proj.weight"] = down_proj
+        else:
+            d[prefix + ".down_proj.weight"] = down_proj
+            d[prefix + ".up_proj.weight"] = up_proj
+
+        d.update(
```

Comment:
i need to finish the branch but this should be very simple once it's merged. You'll have to do your lowrank approximation manually but otherwise all of this will be handled for you.

In case you're curious/want to provide feedback, the branches are:

* https://github.com/stanford-crfm/haliax/pull/90
* https://github.com/stanford-crfm/levanter/pull/608/files

Reply:
Oh that's much nicer indeed. I like the separation of determining the saveable state and then transforming it to HF, and having `Stacked` et. al handle more of the load.

The linearization and `out_first` logic was definitely tedious to work with when I was trying to get this working. I ended up creating one of those weird models where every dimension is a prime number to figure out what was going where :). (It didn't help that I couldn't figure out the real shape that SVD was outputting...)